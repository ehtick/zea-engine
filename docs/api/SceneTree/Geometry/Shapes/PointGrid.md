<a name="PointGrid"></a>

## PointGrid ⇐ <code>Points</code>
A class for generating a point grid.

**Kind**: global class  
**Extends**: <code>Points</code>  

* [PointGrid ⇐ <code>Points</code>](#PointGrid)
    * [new PointGrid(x, y, xDivisions, yDivisions, addTextureCoords)](#new-PointGrid)
    * [x ⇒ <code>number</code>](#x)
    * [x](#x)
    * [y ⇒ <code>number</code>](#y)
    * [y](#y)
    * [getX() ⇒ <code>number</code>](#getX)
    * [setX(val)](#setX)
    * [getY() ⇒ <code>number</code>](#getY)
    * [setY(val)](#setY)
    * [setSize(x, y)](#setSize)
    * [toJSON() ⇒ <code>object</code>](#toJSON)

<a name="new_PointGrid_new"></a>

### new PointGrid
Create a point grid.


| Param | Type | Default | Description |
| --- | --- | --- | --- |
| x | <code>number</code> | <code>1</code> | The length of the point grid along the X axis. |
| y | <code>number</code> | <code>1</code> | The length of the point grid along the Y axis. |
| xDivisions | <code>number</code> | <code>1</code> | The number of divisions along the X axis. |
| yDivisions | <code>number</code> | <code>1</code> | The number of divisions along the Y axis. |
| addTextureCoords | <code>boolean</code> | <code>false</code> | The addTextureCoords value. |

<a name="PointGrid+x"></a>

### x 
Getter for X.Is deprectated. Please use "getX".

**Kind**: instance property of [<code>PointGrid</code>](#PointGrid)  
**Returns**: <code>number</code> - - Returns the length.  
<a name="PointGrid+x"></a>

### pointGrid
Setter for X.Is deprectated. Please use "setX".

**Kind**: instance property of [<code>PointGrid</code>](#PointGrid)  

| Param | Type | Description |
| --- | --- | --- |
| val | <code>number</code> | The length along the X axis. |

<a name="PointGrid+y"></a>

### y 
Getter for Y.Is deprectated. Please use "getY".

**Kind**: instance property of [<code>PointGrid</code>](#PointGrid)  
**Returns**: <code>number</code> - - Returns the length.  
<a name="PointGrid+y"></a>

### pointGrid
Setter for Y.Is deprectated. Please use "setY".

**Kind**: instance property of [<code>PointGrid</code>](#PointGrid)  

| Param | Type | Description |
| --- | --- | --- |
| val | <code>number</code> | The length along the Y axis. |

<a name="PointGrid+getX"></a>

### getX
Getter for the length of the point grid along the X axis.

**Kind**: instance method of [<code>PointGrid</code>](#PointGrid)  
**Returns**: <code>number</code> - - Returns the length.  
<a name="PointGrid+setX"></a>

### setX
Setter for the length of the point grid along the X axis.

**Kind**: instance method of [<code>PointGrid</code>](#PointGrid)  

| Param | Type | Description |
| --- | --- | --- |
| val | <code>number</code> | The length along the X axis. |

<a name="PointGrid+getY"></a>

### getY
Getter for the length of the point grid along the Y axis.

**Kind**: instance method of [<code>PointGrid</code>](#PointGrid)  
**Returns**: <code>number</code> - - Returns the length.  
<a name="PointGrid+setY"></a>

### setY
Setter for the length of the point grid along the Y axis.

**Kind**: instance method of [<code>PointGrid</code>](#PointGrid)  

| Param | Type | Description |
| --- | --- | --- |
| val | <code>number</code> | The length along the Y axis. |

<a name="PointGrid+setSize"></a>

### setSize
Setter for the size of the point grid.

**Kind**: instance method of [<code>PointGrid</code>](#PointGrid)  

| Param | Type | Description |
| --- | --- | --- |
| x | <code>number</code> | The length along the X axis. |
| y | <code>number</code> | The length along the Y axis. |

<a name="PointGrid+toJSON"></a>

### toJSON
The toJSON method encodes this type as a json object for persistences.

**Kind**: instance method of [<code>PointGrid</code>](#PointGrid)  
**Returns**: <code>object</code> - - Returns the json object.  
