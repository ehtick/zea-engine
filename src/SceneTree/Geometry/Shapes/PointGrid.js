import { Vec2 } from '../../../Math/Vec2';
import { Vec3 } from '../../../Math/Vec3';
import { Points } from '../Points.js';

class PointGrid extends Points {
    constructor(x = 1.0, y = 1.0, xDivisions = 1, yDivisions = 1, addTextureCoords=false) {
        super();

        if(isNaN(x) || isNaN(y) || isNaN(xDivisions) || isNaN(yDivisions))
            throw("Invalid geom args");
        
        this.__x = x;
        this.__y = y;
        this.__xDivisions = xDivisions;
        this.__yDivisions = yDivisions;
        if(addTextureCoords)
            this.addVertexAttribute('texCoords', Vec2);
        this.__rebuild();
    }

    get x() {
        console.warn(("getter is deprectated. Please use 'getX'"));
        return this.getX();
    }

    set x(val) {
        console.warn(("getter is deprectated. Please use 'setX'"));
        this.setX(val);
    }

    get y() {
        console.warn(("getter is deprectated. Please use 'getY'"));
        return this.getY();
    }

    set y(val) {
        console.warn(("getter is deprectated. Please use 'setY'"));
        this.setY(val);
    }

    getX() {
        return this.__x
    }

    setX(val) {
        this.__x = val;
        this.__resize();
    }

    getY() {
        return this.__y
    }

    setY(val) {
        this.__y = val;
        this.__resize();
    }

    setSize(x, y) {
        this.__x = x;
        this.__y = y;
        this.__resize();
    }

    __rebuild() {
        this.setNumVertices(this.__xDivisions * this.__yDivisions);

        const texCoords = this.getVertexAttribute('texCoords');
        if(texCoords){
            for (let i = 0; i < this.__yDivisions; i++) {
                let y = i / (this.__yDivisions-1);
                for (let j = 0; j < this.__xDivisions; j++) {
                    let x = j / (this.__xDivisions-1);
                    texCoords.getValueRef((i*this.__xDivisions) + j).set(x, y);
                }
            }
        }

        this.__resize();
    }

    __resize() {
        for (let i = 0; i < this.__yDivisions; i++) {
            let y = ((i / (this.__yDivisions-1)) - 0.5) * this.__y;
            for (let j = 0; j < this.__xDivisions; j++) {
                let x = ((j / (this.__xDivisions-1)) - 0.5) * this.__x;
                this.getVertex((i*this.__xDivisions) + j).set(x, y, 0.0);
            }
        }
        this.setBoundingBoxDirty();
    }

    toJSON() {
        let json = super.toJSON();
        json['x'] = this.__x;
        json['y'] = this.__y;
        json['xDivisions'] = this.__xDivisions;
        json['yDivisions'] = this.__yDivisions;
        return json
    }
};

export {
    PointGrid
};
//export default PointGrid;