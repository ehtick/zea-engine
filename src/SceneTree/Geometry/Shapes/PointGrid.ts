import { ProceduralPoints } from './ProceduralPoints'
import { NumberParameter } from '../../Parameters/NumberParameter'
import { Registry } from '../../../Registry'
import { Vec3Attribute } from '../Vec3Attribute'
import { Vec3 } from '../../../Math'

/**
 * Represents an ordered grid of points along `X` and `Y` axes.
 *
 * ```
 * const pointGrid = new PointGrid(2.2, 1.5, 12, 12)
 * ```
 *
 * **Parameters**
 * * **X(`NumberParameter`):** Length of the grid along the `X` axis.
 * * **Y(`NumberParameter`):** Length of the grid along the `Y` axis.
 * * **XDivisions(`NumberParameter`):** Number of divisions along `X` axis
 * * **YDivisions(`NumberParameter`):** Number of divisions along `Y` axis
 * @extends {ProceduralPoints}
 */
class PointGrid extends ProceduralPoints {
  /**
   * @member sizeXParam - Length of the grid along the `X` axis.
   */
  sizeXParam: NumberParameter

  /**
   * @member divisionsXParam - Number of divisions along `X` axis
   */
  divisionsXParam: NumberParameter

  /**
   * @member sizeYParam - Length of the grid along the `Y` axis.
   */
  sizeYParam: NumberParameter

  /**
   * @member divisionsYParam - Number of divisions along `Y` axis
   */
  divisionsYParam: NumberParameter
  // topologyParams: string[]

  /**
   * Creates an instance of PointGrid.
   *
   * @param x - The length of the point grid along the X axis.
   * @param y - The length of the point grid along the Y axis.
   * @param xDivisions - The number of divisions along the X axis.
   * @param yDivisions - The number of divisions along the Y axis.
   * @param addTextureCoords - The addTextureCoords value.
   */
  constructor(x = 1.0, y = 1.0, xDivisions = 1, yDivisions = 1, addTextureCoords = false) {
    super()
    this.topologyParams = []
    if (isNaN(x) || isNaN(y) || isNaN(xDivisions) || isNaN(yDivisions)) throw new Error('Invalid geom args')
    this.sizeXParam = this.addParameter(new NumberParameter('X', x)) as NumberParameter
    this.sizeYParam = this.addParameter(new NumberParameter('Y', y)) as NumberParameter
    this.divisionsXParam = this.addParameter(new NumberParameter('XDivisions', xDivisions)) as NumberParameter
    this.divisionsYParam = this.addParameter(new NumberParameter('YDivisions', yDivisions)) as NumberParameter

    this.topologyParams.push('XDivisions')
    this.topologyParams.push('YDivisions')
  }

  /**
   * The rebuild method.
   * @private
   */
  rebuild(): void {
    const xDivisions = this.divisionsXParam.value
    const yDivisions = this.divisionsYParam.value
    this.setNumVertices(xDivisions * yDivisions)
    this.resize()
  }

  /**
   * The resize method.
   * @private
   */
  resize(): void {
    const xDivisions = this.divisionsXParam.value
    const yDivisions = this.divisionsYParam.value
    const x = this.sizeXParam.value
    const y = this.sizeYParam.value

    const positions = <Vec3Attribute>this.getVertexAttribute('positions')
    if (!positions) return
    for (let i = 0; i < yDivisions; i++) {
      const newY = (i / (yDivisions - 1) - 0.5) * y
      for (let j = 0; j < xDivisions; j++) {
        const newX = (j / (xDivisions - 1) - 0.5) * x
        positions.setValue(i * xDivisions + j, new Vec3(newX, newY, 0.0))
      }
    }
  }
}

Registry.register('PointGrid', PointGrid)
export { PointGrid }
