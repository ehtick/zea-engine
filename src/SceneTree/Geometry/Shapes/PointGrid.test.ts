import { PointGrid } from './PointGrid'

describe('PointGrid', () => {
  it('tests default parameters', () => {
    const pointGrid = new PointGrid()

    expect(pointGrid.sizeXParam.value).toBe(1)
    expect(pointGrid.sizeYParam.value).toBe(1)
    expect(pointGrid.divisionsXParam.value).toBe(1)
    expect(pointGrid.divisionsYParam.value).toBe(1)
  })

  it('updates parameters', () => {
    const pointGrid = new PointGrid()

    pointGrid.sizeXParam.value = 5
    pointGrid.sizeYParam.value = 5
    pointGrid.divisionsXParam.value = 10
    pointGrid.divisionsYParam.value = 10

    expect(pointGrid.sizeXParam.value).toBe(5)
    expect(pointGrid.sizeYParam.value).toBe(5)
    expect(pointGrid.divisionsXParam.value).toBe(10)
    expect(pointGrid.divisionsYParam.value).toBe(10)
  })

  // There's an issue with flags, that are preventing parameters to be exported.
  it('saves to JSON (serialization).', () => {
    const pointGrid = new PointGrid(3, 3, 6, 6, true)
    const outputJSON = pointGrid.toJSON()

    expect(outputJSON).toMatchSnapshot()
  })

  it('restores from JSON (serialization).', () => {
    const pointGrid = new PointGrid()
    const inputJSON = {
      params: {
        X: {
          value: 3,
        },
        XDivisions: {
          value: 6,
        },
        Y: {
          value: 3,
        },
        YDivisions: {
          value: 6,
        },
      },
      type: 'PointGrid',
      vertexAttributes: {},
    }
    pointGrid.fromJSON(inputJSON)

    const newPointGrid = new PointGrid(3, 3, 6, 6, true)
    expect(pointGrid.toJSON()).toEqual(newPointGrid.toJSON())
  })
})
