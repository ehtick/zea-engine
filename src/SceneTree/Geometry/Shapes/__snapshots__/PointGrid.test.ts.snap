// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PointGrid saves to JSON (serialization). 1`] = `
Object {
  "params": Object {
    "X": Object {
      "value": 3,
    },
    "XDivisions": Object {
      "value": 6,
    },
    "Y": Object {
      "value": 3,
    },
    "YDivisions": Object {
      "value": 6,
    },
  },
  "type": "PointGrid",
  "vertexAttributes": Object {},
}
`;
